VMware: vmware_tools connection plugin

jamescassell opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->
I've updated the referenced PR 
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
Fixes: #47072

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Feature Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
vmware_tools

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->
> > At one point there was a wiki page attempting to document and standardize these going forward, but I'm not sure where that is anymore. I think the only things we all agree on are `ansible_user` and `ansible_password` defaulting for the vm username and password. smile
> 
> Do you mean this - https://github.com/ansible/community/wiki/VMware:-standardization
> 

I've integrated the changes from TLS standardization and connection_var standardization.  I also added support for the VI_* env vars standardized by vmware sdk.

> > I could use some help getting started or some examples of what this needs to look like. Is there a good reference for this in the existing `ansible` codebase?
> 
> I am sure that we can do integration test as govcsim does not emulate actual OS. So unit test will be more than sufficient for this change. You may want to follow existing connection plugin testcases - https://github.com/ansible/ansible/tree/devel/test/units/plugins/connection
> 

I'm punting this for now.  I'd want to do an integration test, but that requires a full VMware setup.

> > > 1. End-user documentation
> > 
> > 
> > I assumed that meant something like the [winrm connection page](https://docs.ansible.com/ansible/latest/plugins/connection/winrm.html), but I thought that would be autogenerated from the `DOCUMENTATION` block. Either way, do we need more?
> 
> Yes, we can add documentation here = https://github.com/ansible/ansible/blob/devel/docs/docsite/rst/vmware/ as how-to-use connection plugin.
> 

I added an EXAMPLES block.

> > I'm not sure how to go about doing that here, it's potentially something different altogether, see the discussion in #43428.
> > I believe the plugin works, as (manually) tested by myself and at least a couple of others in this thread. I think we just need to decide what it needs for the 2.8 release. I'm willing to finish it up, but I need use some clearer definitions of what _finished_ looks like.
> 
> Not sure about this. But I think if we can get initial version in tree, we can always improvise things. Its better to have something than nothing. :)

This PR intends to show what 'finished' looks like.

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

dericcrago committed:
vmware_tools connection plugin

jamescassell committed:
variable standardization and cleanup

- connection_address -> vmware_host

- connection_username -> vmware_user

- connection_password -> vmware_password

- connection_verify_ssl -> validate_certs

- connection_ignore_ssl_warnings -> silence_tls_warnings

- ansible_vmware_tools_vm_path -> ansible_vmware_guest_path

- standardize user/pass vars

- fallback to default ansible conneciton vars

- accept VMware standard env vars:
https://code.vmware.com/docs/6536/vsphere-sdk-for-perl-programming-guide/doc/GUID-66555F76-570A-4C76-BEA7-6C371BF685D6.html

- note lack of "become" support

- add example usage

- more reasonable default sleep interval

- auto-silence tls warnings if validate_certs=false

- get_option for executable

- remove unsafe 'makedirs_safe'

jamescassell commented:
@Akasurde @dericcrago @pdellaert @Im0 @tchernomax @dagwieers @jborean93 @nitzmahone @verdel PTAL.  This is a cleanup of #47072 addressing most of the feedback.

Akasurde commented:
@pdellaert @Im0 @ckotte @jeking3 @Tomorrow9 @alongchamps @adarobin @goneri  Could you please review this ?

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+DOCUMENTATION = """
+    author: Deric Crago <deric.crago@gmail.com>
+    connection: vmware_tools
+    short_description: Execute modules via VMware Tools.
```
```suggestion
    short_description: Execute tasks inside a VM via VMware Tools
```

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+    connection: vmware_tools
+    short_description: Execute modules via VMware Tools.
+    description:
+      - Execute modules via VMware Tools.
```
```suggestion
      - Use VMware tools to run tasks in, or put/fetch files to VM running in VMware infrastructure.
```

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+    short_description: Execute modules via VMware Tools.
+    description:
+      - Execute modules via VMware Tools.
+      - "Note: Windows VMs will need to have C(ansible_shell_type: powershell) set."
```
```suggestion
      - In case of Windows VMs, set C(ansible_shell_type) to C(powershell).
```
I am OK with current description as well.

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+    description:
+      - Execute modules via VMware Tools.
+      - "Note: Windows VMs will need to have C(ansible_shell_type: powershell) set."
+      - Does not work with 'become'
```
```suggestion
      - Does not work with 'become'.
```

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+      - Does not work with 'become'
+    version_added: "2.8"
+    requirements:
+      - pyvmomi (python library)
```
```suggestion
      - pyvmomi (Python library)
```

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+    version_added: "2.8"
+    requirements:
+      - pyvmomi (python library)
+      - requests (python library)
```
```suggestion
      - requests (Python library)
```

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+    options:
+      vmware_host:
+        description:
+          - Address for the connection (vCenter or ESXi Host).
```
```suggestion
          - FQDN or IP Address for the connection (vCenter or ESXi Host).
```

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+        description:
+          - Address for the connection (vCenter or ESXi Host).
+        env:
+          - name: VI_SERVER
```
```suggestion
          - name: VMWARE_HOST
```
This will match [existing environment variable](https://github.com/ansible/ansible/blob/a2eb227970b5f53a648af3daf58d745d2f33067e/lib/ansible/plugins/doc_fragments/vmware.py#L57) used all VMware modules.

jamescassell replied:
I'd like to support both if you don't mind.  Is VMWARE_HOST defined anywhere outside of ansible?

Akasurde replied:
`VMWARE_HOST` is environment variable since inception of VMware_* modules. Even, I don't mind supporting both.

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+               - VirtualMachine.GuestOperations.Modify
+               - VirtualMachine.GuestOperations.Query"
+        env:
+          - name: VI_USERNAME
```
```suggestion
          - name: VMWARE_USER
```
Same as VMWARE_HOST - https://github.com/ansible/ansible/blob/a2eb227970b5f53a648af3daf58d745d2f33067e/lib/ansible/plugins/doc_fragments/vmware.py#L63

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+        description:
+          - Password for the connection.
+        env:
+          - name: VI_PASSWORD
```
```suggestion
          - name: VMWARE_PASSWORD
```

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+        description:
+          - Verify SSL for the connection.
+          - "Note: This will validate certs for both C(vmware_host) and the ESXi host running the VM."
+        vars:
```
```suggestion
        env:
          - name: VMWARE_VALIDATE_CERTS
        vars:
```

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+        type: bool
+      vm_path:
+        description:
+          - VM path relative to the connection.
```
```suggestion
          - VM path absolute to the connection.
```

jamescassell replied:
Might need better wording, but ignoring for now.  I accepted your change on this item.

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+      vm_path:
+        description:
+          - VM path relative to the connection.
+          - "vCenter Example: C(Datacenter/vm/Discovered virtual machine/testVM) (Needs to include C(vm) between the Datacenter and the rest of the VM path.)"
```
```suggestion
          - "vCenter Example: C(Datacenter/vm/Discovered virtual machine/testVM)."
```

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+        description:
+          - VM path relative to the connection.
+          - "vCenter Example: C(Datacenter/vm/Discovered virtual machine/testVM) (Needs to include C(vm) between the Datacenter and the rest of the VM path.)"
+          - "ESXi Host Example: C(ha-datacenter/vm/testVM) (Needs to include C(vm) between the Datacenter (C(ha-datacenter)) and the rest of the VM path.)"
```
```suggestion
          - "ESXi Host Example: C(ha-datacenter/vm/testVM)."
```

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+          - VM path relative to the connection.
+          - "vCenter Example: C(Datacenter/vm/Discovered virtual machine/testVM) (Needs to include C(vm) between the Datacenter and the rest of the VM path.)"
+          - "ESXi Host Example: C(ha-datacenter/vm/testVM) (Needs to include C(vm) between the Datacenter (C(ha-datacenter)) and the rest of the VM path.)"
+          - Must include VM name, appended to 'folder' as would be passed to vmware_guest.
```
```suggestion
          - Must include VM name, appended to 'folder' as would be passed to M(vmware_guest).
          - Needs to include C(vm) between the Datacenter and the rest of the VM path.
          - Datacenter default value for ESXi server is C(ha-datacenter).
          - C(vm) is not visible in vSphere Web Client but necessary for VMware API to work.
```

jamescassell replied:
thanks for the extra detail!

Akasurde replied:
No problem.

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+        required: True
+      vm_user:
+        description:
+          - VM username.
```
```suggestion
          - Username in guest operating system.
```

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+        required: True
+      vm_password:
+        description:
+          - VM password.
```
```suggestion
          - Password for the user in guest operating system.
```

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+        required: True
+      exec_command_sleep_interval:
+        description:
+          - exec command sleep interval in seconds.
```
```suggestion
          - Time in seconds to sleep between execution of command.
```

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+ansible_vmware_host: vcenter.example.com
+ansible_vmware_user: administrator@vsphere.local
+ansible_vmware_password: Secr3tP4ssw0rd!12
+ansible_vmware_validate_certs: no  # default is yes
```
```suggestion
ansible_vmware_validate_certs: False # default is True
```

jamescassell replied:
why? (no big deal, just curious)

Akasurde replied:
Nothing specific, wanted to match other module style.

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+from ssl import SSLEOFError, SSLError
+from time import sleep
+
+import requests
```
```suggestion
import traceback

REQUESTS_IMP_ERR = None
try:
    import requests
    HAS_REQUESTS = True
except ImportError:
    REQUESTS_IMP_ERR = traceback.format_exc()
    HAS_REQUESTS = False
```

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+        vars:
+          - name: ansible_vmware_tools_file_chunk_size
+        default: 128
+        type: integer
```
Also add parameter `VMWARE_PORT`. Sometime vCenter is located on custom port rather than `443`. 

https://github.com/ansible/ansible/blob/a2eb227970b5f53a648af3daf58d745d2f33067e/lib/ansible/plugins/doc_fragments/vmware.py#L85

jamescassell replied:
Yeah, would be good to add... do you immediately know how it would be passed?

Akasurde replied:
It is available in PR. Added on https://github.com/ansible/ansible/pull/55059/files#diff-2c15779308a54e8ac33b5eba806f6e5bR84

jamescassell replied:
nevermind, I see that you added it

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+    def _connect(self):
+        if not HAS_PYVMOMI:
+            raise AnsibleError("missing 'pyvmomi' or dependencies: %s" % to_native(PYVMOMI_IMPORT_ERROR))
+
```
Add same import error block for requests.

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+
+        pid_info = self._get_pid_info(pid)
+
+        while pid_info.endTime is None:
```
how about - 

https://github.com/ansible/ansible/blob/a2eb227970b5f53a648af3daf58d745d2f33067e/lib/ansible/modules/cloud/vmware/vmware_vm_shell.py#L303

Akasurde replied:
Not blocker for now.

Akasurde commented:
@jamescassell Thanks for working on this. Let me know if you need any help with this.

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+
+# example playbook_linux.yml
+---
+- name: Test VMware Tools Connection Plugin
```
```suggestion
- name: Test VMware Tools Connection Plugin for Linux
```

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,452 @@
+
+# example playbook_windows.yml
+---
+- name: Test VMware Tools Connection Plugin
```
```suggestion
- name: Test VMware Tools Connection Plugin for Windows
```

Akasurde committed:
vmware_tools: review comments

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

Akasurde commented:
@jamescassell I added changes here - https://github.com/jamescassell/ansible/pull/1 feel free to modify.

jamescassell commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,482 @@
+
+    def _establish_connection(self):
+        connection_kwargs = {
+            "host": self.vmware_host,
```
I assume it's fine to use get_option here, too?

jamescassell replied:
Seems to have been fine... I made the change and tested that things still work.

jamescassell replied:
okay, I now see it was an '@ property' so the get_option was elsewhere

jamescassell commented:
@Akasurde I added your changes, re-added the VI_* env vars, and then squashed to one commit per contibutor so ansibot will be happy to merge it once approved.

bcoca commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,486 @@
+            program_path = "cmd.exe"
+            arguments = "/c %s" % cmd_parts
+        else:
+            program_path = self._play_context.executable
```
dont use play_context, add 'executable' option to docs and `self.get_option('executable')`

jamescassell replied:
What should be the default value?

Akasurde replied:
I think @bcoca meant something like this - https://github.com/ansible/ansible/blob/166a33ef4d7272ed83c0b365545ccaccc50dd6d5/lib/ansible/plugins/connection/lxd.py#L23

jamescassell replied:
Done.

bcoca commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,486 @@
+    HAS_REQUESTS = True
+except ImportError:
+    REQUESTS_IMP_ERR = traceback.format_exc()
+    HAS_REQUESTS = False
```
i would avoid using requests, introduced dep that is not needed, as functions in urls.py  should cover what this plugin uses.

jamescassell replied:
I agree in principle that the built-in methods should be sufficient, as it's doing one `requests.get` and one `requests.put` ...but I don't know how to translate between the two...

On further inspection, it looks like the built-in urls.py doesn't support chunk_size and piecemeal fetching of remote files, so I'm inclined to just leave the requests dep as-is.

This is only a dep for the controller, not the managed nodes, and all other vmware components in ansible also require python requests

gundalow replied:
Can you please raise an issue for this so we can get the built in updated

bcoca replied:
if its already a pvimomi dependency its ok to use, but then you really dont need to manage it this way, just import it in same block you do pyvimomi and only deal with missing pyvimomi

jamescassell replied:
I'll raise an issue.  On my RHEL7 system, requests doesn't seem to be a direct dep of pyvmomi, but on my Fedora 29 system, python3-requests is a direct dep of python3-pyvmomi.

jamescassell replied:
Issue opened: https://github.com/ansible/ansible/issues/55162

jamescassell commented:
@Akasurde I've integrated the last two of your suggestions.  (Shippable seems broken, however... which seems unrelated)

bcoca commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,486 @@
+          - name: ansible_vmware_validate_certs
+        default: True
+        type: bool
+      silence_tls_warnings:
```
how are these warnings and not failures? is this controlling behaviour if `validate_certs` is False? would we not just ignore warnings then?

jamescassell replied:
It was added by the original author.  @dericcrago is it okay to remove this option and have it default to 'true'?

Akasurde replied:
I think we can set it to True by default so that we can just ignore warnings as user set `validate_certs` to `False`

jamescassell replied:
Now defaults to `true`

bcoca commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,486 @@
+        try:
+            pid = self.processManager.StartProgramInGuest(vm=self.vm, auth=self.vm_auth, spec=guest_program_spec)
+        except vim.fault.NoPermission as e:
+            raise AnsibleError("No Permission Error: %s %s" % (to_native(e.msg), to_native(e.privilegeId)))
```
should these be AnsibleConnection errors instead?

jamescassell replied:
I don't know, but I can change it if it should be.

jamescassell replied:
Done.

Akasurde commented:
@jamescassell Would you be able to address review comments ? Thursday is feature freeze for 2.8. Let me know. Thanks.

jamescassell commented:
> @jamescassell Would you be able to address review comments ? Thursday is feature freeze for 2.8. Let me know. Thanks.

Looks like it boils down to:
- adding `self.get_option('executable')` preferably with a `ansible_vmware_tools_executable` var for configuration
- dropping `silence_tls_warnings`
- AnsibleError -> AnsibleConnection error
- swap python requests for Request from module_utils/urls.py

I can likely get the first 3, but I need to know what the default value should be for `executable`

I'm not sure if I can just drop/replace `requests.get` with `Requests.get` ?

I won't be able to test any changes I make until 7 hours from now.  (I have, of course, tested the PR as-is.)

Akasurde commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,486 @@
+        """Fetch file."""
+        super(Connection, self).fetch_file(in_path, out_path)
+
+        makedirs_safe(dirname(out_path))
```
@jamescassell I think we should remove this as makedirs_safe is not safe. We should not create any directory as user should ensure creation of directory rather than creating them silently.

abadger replied:
s/user/caller/  (although user creation really is the only safe method if the file is in a world-writable directory).

Akasurde replied:
Yup. Totally agreed.

jamescassell replied:
Done.

goneri commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,491 @@
+      executable:
+        description:
+            - shell to use for execution inside container
+        default: /bin/sh
```
Just a suggestion. We can rely on the GuestID here to set the default value: If the GuestId starts with "win" -> PowerShell, otherwise, /bin/sh.

jamescassell replied:
I think this is handled by setting ansible_shell_type=powershell

jamescassell replied:
The plugin already hardcodes it to "cmd.exe" if the guestfamily is windows.

jamescassell commented:
> > @jamescassell Would you be able to address review comments ? Thursday is feature freeze for 2.8. Let me know. Thanks.
> 
> Looks like it boils down to:
> 
> * adding `self.get_option('executable')` preferably with a `ansible_vmware_tools_executable` var for configuration

Done.

> * dropping `silence_tls_warnings`

Changed default to `true` as some thing it's useful.

> * AnsibleError -> AnsibleConnection error

Done.

> * swap python requests for Request from module_utils/urls.py
> 

Will not do, as urls.py doesn't support chunked downloads and the rest of vmware in ansible also requires requests.



Otherwise, I think I've addressed all feedback.

ready_for_review

felixfontein commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,491 @@
+ansible_vmware_user: administrator@vsphere.local
+ansible_vmware_password: Secr3tP4ssw0rd!12
+ansible_vmware_validate_certs: no  # default is yes
+ansible_vmware_silence_tls_warnings: yes # default is no
```
This isn't true anymore.

Akasurde replied:
Good catch.

Akasurde committed:
review comments

Signed-off-by: Abhijeet Kasurde <akasurde@redhat.com>

bcoca commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,488 @@
+        default: /bin/sh
+        vars:
+            - name: ansible_executable
+            - name: ansible_vmware_tools_executable
```
missing ini/env var entries (to use global default)

jamescassell replied:
Added.

bcoca commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,488 @@
+DOCUMENTATION = """
+    author: Deric Crago <deric.crago@gmail.com>
+    connection: vmware_tools
+    short_description: Execute tasks inside a VM via VMware Tools
```
i would mention this plugin is Windows specific

jamescassell replied:
Plugin works for both linux and windows.

bcoca commented on lib/ansible/plugins/connection/vmware_tools.py:
```diff
@@ -0,0 +1,488 @@
+        super(Connection, self).__init__(*args, **kwargs)
+        if hasattr(self, "_shell") and self._shell.SHELL_FAMILY == "powershell":
+            self.module_implementation_preferences = (".ps1", ".exe", "")
+            self.become_methods = ["runas"]
```
these settings make this a 'windows only' connection , but examples above suggest otherwise, which is it?

jamescassell replied:
It supports both.  If SHELL_FAMILY=powershell, it does this special magic.  If not, it "just works" for Linux otherwise.

jamescassell committed:
executable: support env vars and ini

jamescassell commented:
Thanks @Akasurde for the fixes!  I've also addressed the fix suggested by @bcoca of also enabling ini and env for the executable setting.

Let me know if I should re-squash to one commit per author.

ready_for_review

jamescassell commented:
@Akasurde thanks for all your help on this and for merging! @dericcrago thanks for writing it to begin with!
